### ISSUE TITLE
Cannot use deriving-via for IO (or any abstract type)?

### ISSUE THREAD

**Issue by kosmikus** [RyanGlScott/ghc#8]
This fails:
```
{-# LANGUAGE DerivingStrategies #-}
{-# LANGUAGE StandaloneDeriving #-}
module TestIO where

class C f where
  c :: f a -> Int

newtype X f a = X (f a)

instance C (X f) where
  c = const 0

deriving via (X IO a) instance C IO
```
With:
```
TestIO.hs:13:1: error:
    • Can't make a derived instance of
        ‘C IO’ with the DerivSpecVia:: X IO a strategy:
        The data constructors of ‘IO’ are not all in scope
          so you cannot derive an instance for it
    • In the stand-alone deriving instance for ‘C IO’
```

I don't think the implementation of `IO` is relevant here. Am I wrong?

**Comment by RyanGlScott** [comment]
I'm not sure, to be honest. `stock` and `newtype` deriving require that the constructors of the datatype for which you're deriving an instance be in scope so as to avoid breaking abstraction. After all, the code that is generated does (conceptually, at least) depend on the internals of constructors being used.

One could make a similar argument for `deriving via`: the code which is generated does, in some sense, depend on the internals of the datatype, as it has to be representationally equal to the `via` type. But it's possible that this is too onerous of a restriction, as the instance you're trying to derive (`deriving via (X IO a) instance C IO`) clearly needs no knowledge of the internals of `IO` whatsoever. So I could be persuaded either way here.

**Comment by RyanGlScott** [comment]
Also, you've uncovered another bug: the pretty-printing in that error message is abysmal:

```
‘C IO’ with the DerivSpecVia:: X IO a strategy:
```

That really should say "`the via strategy`". This should be straightforward to fix.

**Comment by kosmikus** [comment]
@RyanGlScott Aren't we checking that the two types are `Coercible` into each other? If so, doesn't that check enough? We would not have a `Coercible` instance between `IO` and its *internal* representation in scope, but here, we're wrapping something *around* `IO`, which should in my opinion be totally fine.

**Comment by RyanGlScott** [comment]
Right. You've convinced me that this check doesn't really make sense for `deriving via`, so I think it would be fine to remove it.

Of course, two types being inter-`Coercible` can depend on whether one's constructors are exposed or not, so you might get a typechecker error if you try to standalone-derive something through `deriving via` for an abstract type. This error might be a tad confusing, but then again, so are other errors you can get through `deriving via`, so that's not exactly a new problem.

**Comment by RyanGlScott** [comment]
Fixed in 655cddbe952c1bf168307be77ad8654dc7c6986d.
